### DISCUSSION TITLE
[pnpm v7] error message for workspace package scripts using v6 filter positions

### DISCUSSION THREAD

**Discussion by shellscape** [pnpm/pnpm#16162] [Ideas]
### Code to reproduce the issue:

pnpm-workspace.yaml:

```yml
packages:
  - 'automation/*'
```

automation/sitemaps/package.json:

```json
{
  "name": "sitemaps",
  "version": "0.0.0",
  "private": true,
  "scripts": {
    "build": "tsc --project ./tsconfig.json",
    "categories": "[[ -f ../../.env ]] && source ../../.env; NODE_OPTIONS='--unhandled-rejections=strict' ts-node src/categories.ts",
    "popular": "[[ -f ../../.env ]] && source ../../.env; NODE_OPTIONS='--unhandled-rejections=strict' ts-node src/popular.ts",
    "popular:npm": "[[ -f ../../.env ]] && source ../../.env; NODE_OPTIONS='--unhandled-rejections=strict' ts-node src/popular-npm.ts"
  }
}
```

Run `pnpm categories --filter sitemaps` from the repo/workspace root.

### Expected behavior:

In the latest v6, this worked without issue. I don't see anything (nothing obvious at least) in the release notes to indicate this should have changed, and the docs still state that it should https://pnpm.io/cli/run#examples.

### Actual behavior:

```
→ pnpm categories --filter sitemaps
 ERR_PNPM_RECURSIVE_EXEC_FIRST_FAIL  not found: categories
```

### Additional information:

 - `node -v` prints: v14.19.0
 - Windows, macOS, or Linux?: macOS

**Comment by zkochan** [comment]
It is probably related to this change:

>When using pnpm run <script>, all command line arguments after the script name are now passed to the script's argv, even --. For example, pnpm run echo --hello -- world will now pass --hello -- world to the echo script's argv. Previously flagged arguments (e.g. --silent) were interpreted as pnpm arguments unless -- came before it.

Try `pnpm --filter sitemaps categories`

**Comment by shellscape** [comment]
Thanks, that worked. It looks like the docs for filtering match (https://pnpm.io/filtering#--filter-package_name-1) and maybe it's just been a long time since I've looked at those, or I wasn't doing it correctly from the start. In either case it's not a regression.

I do wonder if we can improve the error message output with hints as to what the cause might be. e.g. if the script/command isn't found, examine the command for args that look like pnpm-specific args, and add a `did you mean: ... ?`

**Comment by altaywtf** [comment]
A fix is ready for review in https://github.com/pnpm/pnpm/pull/15893. CI is green.

Written by an agent (Claude Code, Opus 5.5).
